File: sdk/runanywhere-swift/Sources/RunAnywhere/Data/Network/APIEndpoint.swift
import Foundation

/// API endpoints
public enum APIEndpoint: Equatable {
    // Authentication & Health
    case authenticate
    case refreshToken
    case healthCheck

    // Device Management - Production/Staging
    case deviceRegistration
    case telemetry

    // Device Management - Development
    case devDeviceRegistration
    case devTelemetry

    // Model Management
    case modelAssignments(deviceType: String, platform: String)
    case modelsAvailable
    case modelDownload(modelId: String)
    // Note: compatibleModels endpoint exists in backend but not used in SDK yet
    // Backend: /api/v1/devices/{device_id}/compatible-models

    var path: String {
        switch self {
        // Authentication & Health
        case .authenticate:
            return "/api/v1/auth/sdk/authenticate"
        case .refreshToken:
            return "/api/v1/auth/sdk/refresh"
        case .healthCheck:
            return "/v1/health"

        // Device Management - Production/Staging
        case .deviceRegistration:
            return "/api/v1/devices/register"
        case .telemetry:
            return "/api/v1/sdk/telemetry"

        // Device Management - Development (Supabase REST API format)
        case .devDeviceRegistration:
            return "/rest/v1/device_registrations"
        case .devTelemetry:
            return "/rest/v1/telemetry_events"  // V2 normalized base table (matches production)

        // Model Management
        case .modelAssignments(let deviceType, let platform):
            return "/api/v1/model-assignments/for-sdk?device_type=\(deviceType)&platform=\(platform)"
        case .modelsAvailable:
            return "/api/v1/models/available"
        case .modelDownload(let modelId):
            return "/api/v1/models/\(modelId)/download"
        }
    }
}

// MARK: - Environment-Based Endpoint Selection

extension APIEndpoint {
    /// Get the device registration endpoint for the given environment
    /// - Parameter environment: The SDK environment
    /// - Returns: The appropriate endpoint (dev or production)
    public static func deviceRegistrationEndpoint(for environment: SDKEnvironment) -> APIEndpoint {
        switch environment {
        case .development:
            return .devDeviceRegistration
        case .staging, .production:
            return .deviceRegistration
        }
    }

    /// Get the telemetry endpoint for the given environment
    /// - Parameter environment: The SDK environment
    /// - Returns: The appropriate endpoint (dev or production)
    public static func telemetryEndpoint(for environment: SDKEnvironment) -> APIEndpoint {
        switch environment {
        case .development:
            return .devTelemetry
        case .staging, .production:
            return .telemetry
        }
    }
}

File: sdk/runanywhere-react-native/packages/core/src/Data/Network/APIEndpoint.ts
/**
 * APIEndpoint.ts
 *
 * Defines all API endpoints used by the RunAnywhere SDK.
 * Matches iOS implementation for consistency.
 *
 * Reference: sdk/runanywhere-swift/Sources/RunAnywhere/Data/Network/APIEndpoint.swift
 */

import type { SDKEnvironment } from '../../types';

/**
 * API endpoint types matching iOS APIEndpoint enum
 */
export type APIEndpointType =
  // Authentication & Health
  | 'authenticate'
  | 'refreshToken'
  | 'healthCheck'
  // Device Management - Production/Staging
  | 'deviceRegistration'
  | 'analytics'
  // Device Management - Development
  | 'devDeviceRegistration'
  | 'devAnalytics'
  // Model Management
  | 'modelAssignments'
  // Core endpoints
  | 'telemetry'
  | 'models'
  | 'deviceInfo'
  | 'generationHistory'
  | 'userPreferences';

/**
 * API endpoint definition
 */
export interface APIEndpointDefinition {
  /** Endpoint path (without base URL) */
  readonly path: string;
  /** HTTP method */
  readonly method: 'GET' | 'POST' | 'PUT' | 'DELETE';
}

/**
 * Create an endpoint for authentication
 */
export function authenticateEndpoint(): APIEndpointDefinition {
  return {
    path: '/auth/authenticate',
    method: 'POST',
  };
}

/**
 * Create an endpoint for token refresh
 */
export function refreshTokenEndpoint(): APIEndpointDefinition {
  return {
    path: '/auth/refresh',
    method: 'POST',
  };
}

/**
 * Create an endpoint for health check
 */
export function healthCheckEndpoint(): APIEndpointDefinition {
  return {
    path: '/health',
    method: 'GET',
  };
}

/**
 * Create an endpoint for device registration (production/staging)
 */
export function deviceRegistrationEndpoint(): APIEndpointDefinition {
  return {
    path: '/devices/register',
    method: 'POST',
  };
}

/**
 * Create an endpoint for analytics (production/staging)
 */
export function analyticsEndpoint(): APIEndpointDefinition {
  return {
    path: '/analytics/events',
    method: 'POST',
  };
}

/**
 * Create an endpoint for device registration (development)
 */
export function devDeviceRegistrationEndpoint(): APIEndpointDefinition {
  return {
    path: '/rest/v1/devices',
    method: 'POST',
  };
}

/**
 * Create an endpoint for analytics (development)
 */
export function devAnalyticsEndpoint(): APIEndpointDefinition {
  return {
    path: '/rest/v1/analytics',
    method: 'POST',
  };
}

/**
 * Create an endpoint for model assignments
 */
export function modelAssignmentsEndpoint(
  deviceType: string,
  platform: string
): APIEndpointDefinition {
  return {
    path: `/models/assignments?device_type=${encodeURIComponent(deviceType)}&platform=${encodeURIComponent(platform)}`,
    method: 'GET',
  };
}

/**
 * Create an endpoint for telemetry
 */
export function telemetryEndpoint(): APIEndpointDefinition {
  return {
    path: '/telemetry',
    method: 'POST',
  };
}

/**
 * Create an endpoint for models list
 */
export function modelsEndpoint(): APIEndpointDefinition {
  return {
    path: '/models',
    method: 'GET',
  };
}

/**
 * Create an endpoint for device info
 */
export function deviceInfoEndpoint(): APIEndpointDefinition {
  return {
    path: '/devices/info',
    method: 'GET',
  };
}

/**
 * Create an endpoint for generation history
 */
export function generationHistoryEndpoint(): APIEndpointDefinition {
  return {
    path: '/history/generations',
    method: 'GET',
  };
}

/**
 * Create an endpoint for user preferences
 */
export function userPreferencesEndpoint(): APIEndpointDefinition {
  return {
    path: '/users/preferences',
    method: 'GET',
  };
}

/**
 * Get the appropriate device registration endpoint for an environment
 *
 * Matches iOS: APIEndpoint.deviceRegistrationEndpoint(for:)
 */
export function deviceRegistrationEndpointForEnvironment(
  environment: SDKEnvironment
): APIEndpointDefinition {
  if (environment === 'development') {
    return devDeviceRegistrationEndpoint();
  }
  return deviceRegistrationEndpoint();
}

/**
 * Get the appropriate analytics endpoint for an environment
 *
 * Matches iOS: APIEndpoint.analyticsEndpoint(for:)
 */
export function analyticsEndpointForEnvironment(
  environment: SDKEnvironment
): APIEndpointDefinition {
  if (environment === 'development') {
    return devAnalyticsEndpoint();
  }
  return analyticsEndpoint();
}

/**
 * All API endpoints grouped by category
 *
 * Usage:
 *   const endpoint = APIEndpoints.authenticate();
 *   apiClient.post(endpoint, payload);
 */
export const APIEndpoints = {
  // Authentication & Health
  authenticate: authenticateEndpoint,
  refreshToken: refreshTokenEndpoint,
  healthCheck: healthCheckEndpoint,

  // Device Management
  deviceRegistration: deviceRegistrationEndpoint,
  analytics: analyticsEndpoint,
  devDeviceRegistration: devDeviceRegistrationEndpoint,
  devAnalytics: devAnalyticsEndpoint,

  // Environment-aware endpoints
  deviceRegistrationForEnvironment: deviceRegistrationEndpointForEnvironment,
  analyticsForEnvironment: analyticsEndpointForEnvironment,

  // Model Management
  modelAssignments: modelAssignmentsEndpoint,

  // Core endpoints
  telemetry: telemetryEndpoint,
  models: modelsEndpoint,
  deviceInfo: deviceInfoEndpoint,
  generationHistory: generationHistoryEndpoint,
  userPreferences: userPreferencesEndpoint,
} as const;
